tests/common.py

```diff
@@ -145,18 +146,58 @@ def post_model_outputs_and_maybe_allow_retries(
     stub: service_pb2_grpc.V2Stub,
     request: service_pb2.PostModelOutputsRequest,
     metadata: Tuple,
+    retryable_output_codes: List[int] = [],
 ):
-    return _retry_on_504_on_non_prod(lambda: stub.PostModelOutputs(request, metadata=metadata))
+    response = _retry_on_504_on_non_prod(stub.PostModelOutputs, request=request, metadata=metadata)
+    # this inspect the individual outputs to see if the code means we should retry prediction (returns if retryable_output_codes not passed)
+    response = _retry_on_unsuccessful_predicts_on_non_prod(
+        stub.PostModelOutputs,
+        request=request,
+        metadata=metadata,
+        response=response,
+        retryable_output_codes=retryable_output_codes,
+    )
+    return response
+
+
+def _retry_on_unsuccessful_predicts_on_non_prod(
+    stub_call, request, metadata, response, retryable_output_codes
+):
+    grpc_base = os.environ.get("CLARIFAI_GRPC_BASE", "api.clarifai.com")
+    if grpc_base == "api.clarifai.com":
+        return response  # only retry in nonprod
+    if response.status == status_code_pb2.SUCCESS:
+        return response  # don't retry if top-level response code is SUCCESS
+    if not retryable_output_codes or all(
+        [out.status.code not in retryable_output_codes for out in response.outputs]
+    ):
+        return response  # return if condition to retry is not met
+    for i in range(1, MAX_PREDICT_ATTEMPTS + 1):
+        retry_inputs_by_id = [
```

Comment:
I don't think we should add all this input logic, especially since it looks like it can ignore certain failure cases. Models can be configured wrong and fail when batch_size is a certain size.

Reply:
I'm not sure what failure cases this would ignore. I think that's the point of providing the retryable output codes to know if we should retry or not. I think it's fine for testing.

Reply:
I mentioned in the comment what failure case this ignored: 
> Models can be configured wrong and fail when batch_size is a certain size.
Since you're altering the inputs, that changes the test.

The point of the retry-able codes is for the whole request.

Reply:
We could also add a check for the details of the status, perhaps? To not unintentionally retry on errors with the same code that happen for different reason.

Reply:
I think that becomes a little hackier, especially since descriptions are more likely to change than the status code. Shouldn't the failure we're trying to catch here be some type of timeout? `INTERNAL_UNCATEGORIZED` does seem to broad

Reply:
For the error in question these are the statuses:

```
output:
    code: INTERNAL_UNCATEGORIZED
    description: "Internal error"
    details: "Backend service had unexpected error. Please contact support@clarifai.com with error code: 99009."
    stack_trace: "./proto/utils/request_meta_utils.go:18"
    stack_trace: "./grpc/media_processor/client.go:185"
    stack_trace: "./api/predictions/model_predictions.go:1491"
    stack_trace: "./api/predictions/model_predictions.go:719"
    stack_trace: "runtime/asm_amd64.s:1594"
    internal_details: ";  Backend Traceback:   File \"/home/ubuntu/work/clarifai/utils/map_rpc.py\", line 79, in _get_results     (status, result) = futures.pop(0).result(timeout=timeout)   File \"/usr/local/lib/python3.9/concurrent/futures/_base.py\", line 448, in result     raise TimeoutError()  ; group id: 246cebeb7c6b4d2293ca59b728770db0"
```

and this is the status of the response:
```
  code: FAILURE
  description: "Failure"
  req_id: "94605dd8410f95e1d65aa26890325c1e"
```

Reply:
Seems like we need to handle the error internally better. This particular issue was a timeout issue, but we're returning INTERNAL_UNCATEGORIZED

Reply:
Created a ticket to update that: https://clarifai.atlassian.net/browse/EAGLE-2662

Should we put this on hold or are we okay with retrying on `INTERNAL_UNCATEGORIZED` for now?

Reply:
The top level status is `FAILURE` so just updated it to retry on that status code.

Reply:
Just tested with a script in staging, retries worked. Should be ready fro re-review